### PR TITLE
task2-components-directives-pipes

### PR DESCRIPTION
**Pull request by Doooodie**
1. Task: https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/components-directives-pipes.md
2. Screenshot:
image
3. Deploy: https://doodie-youtube.netlify.app
4. Done 15.01.2024 / no deadline
5. Score: 100 / 100
- [x] Main layout is implemented (**+10**)
- [x] Fake search functionality is implemented: by submitting the form in the **Header** component, search results are shown (**+20**)
- [x] The **Search result item** component contains all the necessary data (**+10**)
- [x] The **Filtering criteria block** toggle functionality is implemented (**+5**)
- [x] The colored border under the **Search result item** is implemented using a `Directive` (**+15**)
- [x] Search result sorting is implemented (both ascending and descending directions) (**+15**)
- [x] Search result filtering by key words is implemented using a `Pipe` (**+15**)
- [x] Custom Button component is implemented and used across the application. The component utilizes [Content projection](https://angular.io/guide/content-projection) to display the button text (**+5**)
- [x] One of the components is implemented as [standalone](https://angular.io/guide/standalone-components). Strong recommendation to choose a simple presentational component and not a "Smart" container component for this task (**+5**)

### CODE REVIEW

File: src/app/header/search-input/search-input.component.html [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<div class="join">
+  <input
+    type="text"
+    class="input join-item input-bordered input-xs sm:input-sm md:input-md"
+    (input)="changeQuery($event)"
```
Comment (sergeantops134) [inline src/app/header/search-input/search-input.component.html:5]: [minor] Looks like Reactive form could be used, but for a single input it is not so crucial

File: src/app/header/filter-input/filter-input.component.html [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<div class="flex gap-1">
+  <span class="text-primary">by word or sentence</span>
+  <input
+    type="text"
+    class="input input-xs"
+    (input)="changeQuery($event)"
```
Comment (sergeantops134) [inline src/app/header/filter-input/filter-input.component.html:6]: [minor] Looks like Reactive form could be used, but for a single input it is not so crucial

File: src/app/search/pipes/filter-results.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    const result = [];
+
+    const filtered = items.filter((item) => {
+      return item.snippet.title.toLowerCase().includes(query) && item;
+    });
+
+    result.push(...filtered);
+
+    return result;
```
Comment (sergeantops134) [inline src/app/search/pipes/filter-results.pipe.ts:20]: [minor] Possible logic optimisation:

`return items.filter(item => item?.snippet.title.toLowerCase().includes(query.toLowerCase));`

File: src/app/search/directives/change-border-color.directive.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  private setClassByDate() {
+    const now = new Date(2019, 9, 29, 16, 0, 14); // use past date for mocked data
+    const week = new Date(1970, 0, 7);
+    const month = new Date(1970, 1);
+    const halfYear = new Date(1970, 6);
+    const publishedAt = new Date(this.appChangeBorderColor);
+    const diff = new Date(+now - +publishedAt);
+
+    if (diff < week) {
+      this.appChangeBorderColor = 'border-info';
```
Comment (sergeantops134) [inline src/app/search/directives/change-border-color.directive.ts:21]: I would suggest to add another property to store class name

File: src/app/shared/services/video-items.service.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  private setSortStatus(sortBy: SortKeys) {
+    if (this.sortStatus()[sortBy] === SortValues.off) {
+      this.sortStatus.set({ ...this.sortStatus(), [sortBy]: SortValues.asc });
+    } else if (this.sortStatus()[sortBy] === SortValues.asc) {
+      this.sortStatus.set({ ...this.sortStatus(), [sortBy]: SortValues.desc });
+    } else if (this.sortStatus()[sortBy] === SortValues.desc) {
+      this.sortStatus.set({ ...this.sortStatus(), [sortBy]: SortValues.off });
+    }
```
Comment (sergeantops134) [inline src/app/shared/services/video-items.service.ts:29]: [minor] Possible logic optimisation


const sortStatusMap = {
 [SortValues.off]: SortValues.asc,
 [SortValues.asc]: SortValues.desc,
 [SortValues.desc]: SortValues.off
};

this.sortStatus.set({
 ...this.sortStatus(),
  [sortBy]: sortStatusMap[this.sortStatus()[sortBy] 
  });

File: src/app/header/search-input/search-input.component.html (hunk 1 above)
Comment (Doooodie) [reply src/app/header/search-input/search-input.component.html:5]: I should use forms in further tasks

**Pull request merged**